### DISCUSSION TITLE
issue with ts jest and node-fetch v3

### DISCUSSION THREAD

**Discussion by a-wf** [node-fetch/node-fetch#1503] [Q&A]
I upgraded node-fetch to v3, and run jest test and I get: 

```
 Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:
import http from 'http';
    ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      1 | import { Task, DataError } from '@koober/std';
    > 2 | import fetch from 'node-fetch';
        | ^
      3 | import { FacebookResponseError, FacebookTokenInfo, FacebookTokenInfoFormatted, FacebookUserIdentity } from './data';
      4 |
      5 | export interface FacebookClient

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1597:14)
      at Object.<anonymous> (src/bundle/facebook/client.ts:2:1)
  ```
  
  I followed instructions, but nothing changes, 
  I also tried  this jest config 
```json
  "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "globals": {
      "ts-jest": {
        "tsconfig": "tsconfig.json",
        "useESM": true
      }
    },
     "preset": "ts-jest/presets/default-esm",
```
    
but nothing happens

**Comment by tomas-hartman** [comment]
+1 same issue with ts/jest compiled with babel

```js
  // jest.config.js option that might be relevant
  transform: {
    '^.+\\.[t|j]sx?$': 'babel-jest',
  },
```

**Comment by lynxtaa** [comment]
I think I made it work with `jest` and Typescript. Run jest with `experimental-vm-modules` enabled:
```json
"test:watch": "node --experimental-vm-modules node_modules/jest/bin/jest.js --watch",
```

### jest.config.json if you are using ES Modules:
```json
{
  "roots": ["<rootDir>/src"],
  "extensionsToTreatAsEsm": [".ts"],
  "testEnvironment": "node",
  "moduleNameMapper": {
    "^(\\.{1,2}/.*)\\.js$": "$1"
  },
  "transform": {
    "\\.[jt]sx?$": [
      "babel-jest",
      {
        "babelrc": false,
        "presets": ["@babel/preset-typescript"],
        "plugins": ["@babel/plugin-proposal-optional-chaining"]
      }
    ]
  }
}
```

Check you my example repository: https://github.com/lynxtaa/node-fetch-jest-example

### jest.config.json if you are using CommonJS:
```json
{
  "roots": ["<rootDir>/src"],
  "testEnvironment": "node",
  "transform": {
    "\\.[jt]sx?$": [
      "babel-jest",
      {
        "babelrc": false,
        "presets": ["@babel/preset-typescript"],
        "plugins": [
          "@babel/plugin-proposal-optional-chaining",
          "@babel/plugin-transform-modules-commonjs"
        ]
      }
    ]
  }
}
```
And in this case you should use dynamic import:
```js
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));
```

**Comment by storyalex** [comment]
I have a few typescript projects where this happened to me.

Add the following to `jest.config.js` to run all .js-files through babel when running jest:
```js
  transform: {
    "^.+\\.(ts|tsx)$": "ts-jest",
    "^.+\\.(js)$": "babel-jest",
  },
  transformIgnorePatterns: [
  ],
```

Note the empty `transformIgnorePatterns`. By default transforms ignore node_modules.
This worked for me with:

node-fetch: ^3.0.0
babel-jest: ^27.3.1
jest: ^27.3.1
ts-jest: ^27.0.7

**Comment by martsie** [comment]
Rolling back to 2.6 worked for me to get serverless-bundle to run jest tests.

**Comment by ccarcaci** [comment]
Same here, rolling back to 2.6 worked. Any of the other solutions were not working.

**Reply by sergiycheck** [reply]
worked with **node-fetch: 2.6.7**
[example project (ts, ts-jest, node-fetch)](https://github.com/sergiycheck/http-testing-app)

**Comment by rguzg** [comment]
As typing information is already included in node-fetch 3.0 and up,
the @types/node-fetch package after version 3.0 is just a stub. To get
typing information you need to install a version of the types package
that's inside the 2.x range. In my case, I installed version 2.5.12 and
everything's working fine.

**Comment by ravshansbox** [comment]
This is a small example where jest works with node-fetch@3: https://github.com/ravshansbox/jest-node-fetch-es-modules

**Reply by yunnysunny** [reply]
> This is a small example where jest works with node-fetch@3: https://github.com/ravshansbox/jest-node-fetch-es-modules

Your link is 404 now.

**Comment by plemelin** [comment]
*Note: edited after original comment* this issue should be closed and reference https://github.com/node-fetch/node-fetch/issues/1279 *End note*

> This is a small example where jest works with node-fetch@3: https://github.com/ravshansbox/jest-node-fetch-es-modules

This is misleading as it works only because you are telling jest to run with support for ESM (--experimental-vm-modules). 

As far as I can tell, node-fetch v3 only works as an ESM and you cannot make it work in jest without that flag. Our teams are not going to use that flag unfortunately. So unless this becomes supported in Jest, we are going to use node-fetch <3.0.0

Also looks like it's slowing the adoption of  node-fetch v3:

![Screenshot from 2022-02-09 15-07-41](https://user-images.githubusercontent.com/38959411/153282564-84027f94-37d0-4b1d-8445-365721512ddf.png)

**Comment by jimmywarting** [comment]
I haven't played with jest lately...
but is the "commonjs" vm able to load esm-only packages using the lazy/async `import('node-fetch')`?

**Comment by vrfdivino** [comment]
Using ```ts-jest``` with ```node-fetch@3``` still won't work. I suggest use the following version packages. Still works great!

```json
    "@types/node-fetch": "2",
    "node-fetch": "2",
```

**Comment by Brian-McBride** [comment]
It is pretty clear that the NodeJS and DOM ecosystem is not ready for 100% ESM modules. 
Locking in a lower version creates confusion. It seems that there could have been a period of supporting both until ESM became more mainstream. 
Dropping all commonjs support seems to be more idealistic than meeting the realistic needs of production.

**Reply by madamiak** [reply]
It's just frustrating when you get a SNYK report with the high severity security issue in 2.6.7 and you need to bump to 3.2.10 to resolve it (https://security.snyk.io/vuln/SNYK-JS-NODEFETCH-2964180), but it doesn't work because of Jest issues. What should we do then?

**Reply by bahrmichael** [reply]
@madamiak If you configure Snyk to only raise issues for dependencies (but not devDependencies), you may decide to only override the dependencies' dependencies. E.g. puppeteer-core has a dependency on node-fetch, which one can override in the package.json as below

```
  "overrides": {
    "puppeteer-core": {
      "node-fetch": "^3.2.10"
    }
  }
```

You can then decide to leave Jest at an older version with the vulnerability in place, but without bundling that vuln into your production code.

**Reply by Brian-McBride** [reply]
I am all for pushing to current releases and standards. 
It is clear that ESM is just a bit "too new" for the ecosystem. 

Patching a ton of existing tools across thousands to millions of repos to ignore the current version of node-fetch is just burning so many man hours in this world. 

I heard the complaint was maintaining two code bases in one repo and the final size. Supporting commonjs is a tooling issue, not at all some manual task every release. And size... We can live with a little bloat during the transitions to ESM across the board.

I very much appreciate the open-source maintainers here. And whatever you all want to do, I'll roll with it even if I don't think it is the best decision for the greater community at this time. I do support the move to ESM only, it is clearly just a little too early at the moment.

**Comment by ruslansadygov** [comment]
steps are next:

1. jest.config.cjs add 
`moduleNameMapper: {
    "node-fetch": "<rootDir>/node_modules/node-fetch-jest"
  }`
2. and package.json dev dependencies
`"node-fetch-jest": "npm:node-fetch@^2.6.9",`
3. package.json dependencies
`"node-fetch": "^3.3.0"`

**Reply by ravshansbox** [reply]
Isn't it the same as requiring `"node-fetch": "^2"` in `package.json`?

**Reply by ruslansadygov** [reply]
nope, because your actual prod code will use latest `node-fetch`, but for test you will use compatible version as here https://jestjs.io/docs/next/bypassing-module-mocks

**Reply by iamrenejr** [reply]
This works for me.

**Reply by rafalzawadzki** [reply]
this is an elegant solution

**Comment by helbing** [comment]
If you use `ts-jest`, this document will help you: 

https://kulshekhar.github.io/ts-jest/docs/guides/esm-support/

And you also need to know one thing, if you use `snapshotResolver` in `jest.config.ts`, it will not work, because  `snapshotResolver` does not support loading ESM. The support of loading ESM is still progressing.

facebook/jest/issues/11167

**Comment by devjiwonchoi** [comment]
@a-wf You can use `transformIgnorePatterns` in jest.config.

```js
transformIgnorePatterns: [
  "/node_modules/(?!node-fetch)/"
]
```

> package.json

```json
"transformIgnorePatterns": [
  "/node_modules/(?!node-fetch)/"
]
```
